### ISSUE TITLE
BadRecordMac when using AES-GCM on ARMv7 (RPi3)

### ISSUE THREAD

**Issue by PAStheLoD** [ctz/rustls#149]
Hello,

Using 0.12, and the client from the examples/ we found that the TLS connection gets aborted with an BadRecordMac Alert. We tried cross-compiling (using japaric's cross) and building "natively" on the target device (Raspberry Pi 3 Model B, using Ubuntu Xenial 16.04). The same code works on x86-64 (amd64).

For example connecting to yahoo.com, reddit.com, apache.org both result in the aforementioned error.

```
DEBUG:<unknown>: Using ciphersuite TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
ERROR:<unknown>: TLS alert received: Message {                                                                                                                                                
    typ: Alert,                                                                                                                                                                               
    version: TLSv1_2,                                                                                                                                                                         
    payload: Alert(                                                                                                                                                                           
        AlertMessagePayload {                                                                                                                                                                 
            level: Fatal,                                                                                                                                                                     
            description: BadRecordMac                                                                                                                                                         
        }                                                                                                                                                                     
    )                                                                                         
}
``` 

A workaround (for Nginx server side) is getting an OpenSSL 1.1-built Nginx and enabling TLSv1.3 and specifying ECDHE_RSA_CHACHA20_POLY1305 in ssl_ciphers.

**Comment by briansmith** [comment]
Can you please post a pcap somewhere?

**Comment by briansmith** [comment]
Also, are you able to run the *ring* and/or Rustls test suites on that hardware?

**Comment by PAStheLoD** [comment]
Sure, please see the PCAP file: [pcap.zip](https://github.com/ctz/rustls/files/1664997/pcap.zip)

And yes, we can run the test suite. Is that simply `cargo test`?

**Comment by ctz** [comment]
cargo test will get you most of the tests, yes.

I have a rpi3 that I can test this on too.

**Comment by briansmith** [comment]
Thanks for sharing the pcap.

If you can, it would probably be even more helpful if we had a PCAP plus a SSLKEYLOGFILE-style log that would allow us to reconstruct the keys from the master key.

**Comment by briansmith** [comment]
On the ring side, any failures running `cargo test` and/or `cargo test --release`?

**Comment by PAStheLoD** [comment]
How to capture SSLKEYLOGFILE with rustls?

I'm currently running the rustls tests. Then I'll do the ring tests.

**Comment by PAStheLoD** [comment]
rustls current master, cargo test:

```
...
test client_stream_read ... ok                                             
test client_stream_write ... ok                                              
error: process didn't exit successfully: `/root/rustls-master/target/debug/deps/api-388a9736ff0c92bd` (signal: 7, SIGBUS: access to undefined memory)
```

ring current master, cargo test --verbose:
```
   Compiling ring v0.13.0-alpha (file:///root/ring-master)                                                                                                                                    
     Running `rustc --crate-name build_script_build build.rs --crate-type bin --emit=dep-info,link -C debuginfo=2 --cfg 'feature="default"' --cfg 'feature="dev_urandom_fallback"' --cfg 'feat
ure="use_heap"' -C metadata=0bae6a18afdf8343 -C extra-filename=-0bae6a18afdf8343 --out-dir /root/ring-master/target/debug/build/ring-0bae6a18afdf8343 -L dependency=/root/ring-master/target/d
ebug/deps --extern cc=/root/ring-master/target/debug/deps/libcc-02e2c61d0fca2128.rlib --extern rayon=/root/ring-master/target/debug/deps/librayon-a85031681c34492d.rlib`                      
     Running `/root/ring-master/target/debug/build/ring-0bae6a18afdf8343/build-script-build`                                                                                                  
error: failed to run custom build command for `ring v0.13.0-alpha (file:///root/ring-master)`                                                                                                 
process didn't exit successfully: `/root/ring-master/target/debug/build/ring-0bae6a18afdf8343/build-script-build` (exit code: 101)     

...

--- stderr                                                                                                                                                                                    
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/aesv8-armx-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:
906:4                                                                                                                                                                                         
note: Run with `RUST_BACKTRACE=1` for a backtrace.                                                                                                                                            
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/ghashv8-armx-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.r
s:906:4                                                                                                                                                                                       
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/aes-armv4-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:9
06:4                                                                                                                                                                                          
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/bsaes-armv7-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs
:906:4                                                                                                                                                                                        
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/armv4-mont-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs:
906:4                                                                                                                                                                                         
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/chacha-armv4-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.r
s:906:4
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/ecp_nistz256-armv4-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/re
sult.rs:906:4
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/ghash-armv4-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.rs
:906:4
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/poly1305-armv4-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result
.rs:906:4
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/sha256-armv4-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.r
s:906:4
thread '<unnamed>' panicked at '"/root/ring-master/pregenerated/sha512-armv4-linux32.S": Error { repr: Os { code: 2, message: "No such file or directory" } }', /checkout/src/libcore/result.r
s:906:4

```

**Comment by ctz** [comment]
Both pass for me on my rpi3.

Is it possible your pi is unwell in some way? (In my experience this kind of thing with a raspberry pi is a power issue or the sdcard dying.)

**Comment by briansmith** [comment]
> ring current master, cargo test --verbose:

i think you're running the *ring* test suite from the master branch outside a directory without a `.git` folder. *ring* won't build that way because it is expecting some of the PerlAsm pregeneration to have been done (those files are stuffed into the crate in a special pre-packaging step). Just create a ".git" directory (or file) where one would normally go in a git checkout and run the tests again. (The build would then require Perl, though.)

**Comment by PAStheLoD** [comment]
@ctz  The Pi seems to be okay, at least there are no kernel oopses or other errors in dmesg. OpenVPN runs fine using AES-256-GCM-SHA384.

`Linux wash-vecses2 4.9.41-v7+ #1023 SMP Tue Aug 8 16:00:15 BST 2017 armv7l armv7l armv7l GNU/Linux`


I haven't installed git, so tests were run after a wget github/ring.zip && unzip master.zip, but now after an `mkdir .git` the output is the following:

```
~/ring-master# cargo test --verbose  
       Fresh either v1.4.0                             
       Fresh scopeguard v0.3.3                             
       Fresh lazy_static v1.0.0                        
       Fresh libc v0.2.36                                    
       Fresh cc v1.0.4                                  
       Fresh untrusted v0.6.1                                               
       Fresh lazy_static v0.2.11                         
       Fresh coco v0.1.1                                                 
       Fresh num_cpus v1.8.0
       Fresh rand v0.3.20
       Fresh rayon-core v1.3.0
       Fresh rayon v0.9.0
       Fresh ring v0.13.0-alpha (file:///root/ring-master)
    Finished dev [unoptimized + debuginfo] target(s) in 0.2 secs
     Running `/root/ring-master/target/debug/deps/ring-e7493204881269f6`

running 83 tests
test aead::aes_gcm::tests::max_input_len_test ... ok
test c::test_i16_metrics ... ok
test c::test_i64_metrics ... ok
test c::test_i8_metrics ... ok
test c::test_int_metrics ... ok
test c::test_long_metrics ... ok
test c::test_size_t_metrics ... ok
test aead::aes_gcm::tests::test_aes ... FAILED
test c::test_u16_metrics ... ok
test aead::chacha20_poly1305::tests::max_input_len_test ... ok
test c::test_u32_metrics ... ok
test c::test_i32_metrics ... ok
test c::test_u64_metrics ... ok
test c::test_u8_metrics ... ok
test c::test_uint_metrics ... ok
...
failures:

---- aead::aes_gcm::tests::test_aes stdout ----
        thread 'aead::aes_gcm::tests::test_aes' panicked at 'assertion failed: `(left == right)`
  left: `[121, 11, 109, 252, 108, 61, 63, 255, 205, 252, 211, 95, 187, 220, 229, 15]`,
 right: `[105, 196, 224, 216, 106, 123, 4, 48, 216, 205, 183, 128, 112, 180, 197, 90]`', src/aead/aes_gcm.rs:151:12
note: Run with `RUST_BACKTRACE=1` for a backtrace.
src/aead/aes_tests.txt: Test panicked.
Key = 000102030405060708090a0b0c0d0e0f
Input = 00112233445566778899aabbccddeeff
Output = 69c4e0d86a7b0430d8cdb78070b4c55a
thread 'aead::aes_gcm::tests::test_aes' panicked at 'assertion failed: `(left == right)`
  left: `[138, 63, 183, 175, 198, 219, 55, 142, 49, 98, 111, 77, 129, 129, 172, 158]`,
 right: `[142, 162, 183, 202, 81, 103, 69, 191, 234, 252, 73, 144, 75, 73, 96, 137]`', src/aead/aes_gcm.rs:151:12
src/aead/aes_tests.txt: Test panicked.
Key = 000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f
Input = 00112233445566778899aabbccddeeff
Output = 8ea2b7ca516745bfeafc49904b496089
thread 'aead::aes_gcm::tests::test_aes' panicked at 'Test failed.', src/test.rs:348:8


failures:
    aead::aes_gcm::tests::test_aes

test result: FAILED. 82 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '--lib'
```

With RUST_BACKTRACE=full:

```
...
   5:   0x4ed803 - std::panicking::begin_panic::h995b0003439546f8
                       at /checkout/src/libstd/panicking.rs:538
   6:   0x50926b - ring::test::from_file::h678addd85f561773
                       at src/test.rs:348
   7:   0x4be7cb - ring::aead::aes_gcm::tests::test_aes::h71d1a3c3c95d2f41
                       at src/aead/aes_gcm.rs:125
   8:   0x53342b - <F as test::FnBox<T>>::call_box::h0d75d89fa03308a9
                       at /checkout/src/libtest/lib.rs:1510
                       at /checkout/src/libcore/ops/function.rs:223
                       at /checkout/src/libtest/lib.rs:142
   9:   0x56b6ff - __rust_maybe_catch_panic
                       at /checkout/src/libpanic_unwind/lib.rs:101
...
```

If you need any details, let me know!


We experienced the same BadRecordMac on other RPis too.

**Comment by ctz** [comment]
Ah, thanks for the confirmation. It's very possible that my pi has an older kernel/toolchain than yours (it was in a cupboard for about a year prior to yesterday). I'll update it and try to reproduce again.

**Comment by PAStheLoD** [comment]
Also regarding the rustls test SIGBUS, dmesg has something informative (after `echo 3 > /proc/cpu/alignment` ... meh):

```
Jan 26 12:19:21 xxx kernel: Alignment trap: negotiated_ciph (2339) PC=0x006d9df8 Instr=0xecd6cb04 Address=0x006d9721 FSR 0x001
Jan 26 12:19:21 xxx kernel: Alignment trap: not handling instruction ecd6cb04 at [<006d9df8>]
Jan 26 12:19:21 xxx kernel: Unhandled fault: alignment exception (0x001) at 0x006d9721
Jan 26 12:19:21 xxx kernel: pgd = 9e2d8000
Jan 26 12:19:22 xxx kernel: [006d9721] *pgd=1e3ce835, *pte=0da5a59f, *ppte=0da5ae7e
Jan 26 12:19:23 xxx kernel: Alignment trap: negotiated_ciph (2339) PC=0x006d9f0a Instr=0xecd80b04 Address=0x006d9731 FSR 0x001
Jan 26 12:19:23 xxx kernel: Alignment trap: not handling instruction ecd80b04 at [<006d9f0a>]
Jan 26 12:19:23 xxx kernel: Unhandled fault: alignment exception (0x001) at 0x006d9731
Jan 26 12:19:23 xxx kernel: pgd = 9e2d8000
Jan 26 12:19:23 xxx kernel: [006d9731] *pgd=1e3ce835, *pte=0da5a59f, *ppte=0da5ae7e
```

/proc/cpu/alignment:
```
User:           4
System:         1226 (__csum_ipv6_magic+0x1c/0xc8)
Skipped:        4
Half:           0
Word:           0
DWord:          0
Multi:          1226
User faults:    3 (fixup+warn)
```

So it should fixup the unaligned access, but it probably can't. And there were no other alignment issues during cargo test.

**Comment by briansmith** [comment]
@PAStheLoD Please file an issue in the *ring* issue tracker at https://github.com/briansmith/ring/issues, since it seems to be an issue with *ring*. I wouldn't be surprised if the OpenSSL code we've inherited is making assumptions about alignment that aren't valid. I'm happy to look into fixing them.

**Comment by PAStheLoD** [comment]
@briansmith Will do. Thanks!

Interestingly, using the old ARMv6 target helps with tests to progess further, but then I get this:

```
 cargo test --target arm-unknown-linux-gnueabihf --test badssl
    Finished dev [unoptimized + debuginfo] target(s) in 0.1 secs
     Running target/arm-unknown-linux-gnueabihf/debug/deps/badssl-be5adcc55b30c6a8
    
...
     Running target/arm-unknown-linux-gnueabihf/debug/deps/badssl-be5adcc55b30c6a8                    
                                                                                               
running 14 tests                                                                                
test online::expired ... ok                                                                 
test online::mozilla_inter ... FAILED                                                                 
test online::mozilla_modern ... FAILED                                                         
test online::many_sans ... FAILED                                          
test online::no_cbc ... ok                                                                  
test online::no_dh ... ok                                                      
test online::mozilla_old ... FAILED                      
test online::no_rc4 ... ok                                      
test online::sha1_2016 ... ok                                               
test online::self_signed ... ok                                             
test online::sha256 ... FAILED                                             
test online::rsa8192 ... FAILED                                                                
test online::wrong_host ... ok                                        
test online::too_many_sans ... ok                                            
                                                                                  
failures:                                                                
                                                         
---- online::mozilla_inter stdout ----                          
        We expected to find '<title>mozilla-intermediate.badssl.com</title>' in the following output:
Output { status: ExitStatus(ExitStatus(256)), stdout: "TLS error: AlertReceived(BadRecordMac)\nConnection closed\n", stderr: "" }
thread 'online::mozilla_inter' panicked at 'Test failed', tests/common/mod.rs:289:16
note: Run with `RUST_BACKTRACE=1` for a backtrace.

---- online::mozilla_modern stdout ----
        We expected to find '<title>mozilla-modern.badssl.com</title>' in the following output:
Output { status: ExitStatus(ExitStatus(256)), stdout: "TLS error: AlertReceived(BadRecordMac)\nConnection closed\n", stderr: "" }
thread 'online::mozilla_modern' panicked at 'Test failed', tests/common/mod.rs:289:16

---- online::many_sans stdout ----
        We expected to find '<title>1000-sans.badssl.com</title>' in the following output:
Output { status: ExitStatus(ExitStatus(256)), stdout: "TLS error: AlertReceived(BadRecordMac)\nConnection closed\n", stderr: "" }
thread 'online::many_sans' panicked at 'Test failed', tests/common/mod.rs:289:16

---- online::mozilla_old stdout ----
        We expected to find '<title>mozilla-old.badssl.com</title>' in the following output:
Output { status: ExitStatus(ExitStatus(256)), stdout: "TLS error: AlertReceived(BadRecordMac)\nConnection closed\n", stderr: "" }
thread 'online::mozilla_old' panicked at 'Test failed', tests/common/mod.rs:289:16

---- online::sha256 stdout ----
        We expected to find '<title>sha256.badssl.com</title>' in the following output:
Output { status: ExitStatus(ExitStatus(256)), stdout: "TLS error: AlertReceived(BadRecordMac)\nConnection closed\n", stderr: "" }
thread 'online::sha256' panicked at 'Test failed', tests/common/mod.rs:289:16

---- online::rsa8192 stdout ----
        We expected to find '<title>rsa8192.badssl.com</title>' in the following output:
Output { status: ExitStatus(ExitStatus(256)), stdout: "TLS error: AlertReceived(BadRecordMac)\nConnection closed\n", stderr: "" }
thread 'online::rsa8192' panicked at 'Test failed', tests/common/mod.rs:289:16


failures:
    online::many_sans
    online::mozilla_inter
    online::mozilla_modern
    online::mozilla_old
    online::rsa8192
    online::sha256

test result: FAILED. 8 passed; 6 failed; 0 ignored; 0 measured; 0 filtered out
```

**Comment by ctz** [comment]
I'm going to close this as being a *ring* issue. Please reopen if there's work to do in rustls.

**Comment by briansmith** [comment]
I believe this issue might have been fixed in *ring* 0.14.2 as described in https://github.com/briansmith/ring/issues/628#issuecomment-455726680.

@PAStheLoD Were you running in 64-bit mode (Aarch64)? @ctz where you running a 32-bit OS on your Raspberry Pi 3? If so, that would explain why you got different results, because the bug was AAarch64-specific.

**Comment by PAStheLoD** [comment]
As far as I know I wasn't. Ubuntu 16.04 on rpi3 is in armv7 mode, and the copy-pasted test output is from an armv6 targeted binary. So unless the specific executable could somehow switch to Aarch64 mode it wasn't running in that mode.
